You see, it triggered again: 

```
E   helpers.client.QueryRuntimeException: Client failed! Return code: -6, stderr: Logical error: 'Received from 172.16.4.8:9000. DB::Exception: The rule Detect passwords in tests was triggered on the log line disk(readonly = 1, type = 's3_plain_rewritable', endpoint = 'http://minio1:9001/root/data/', access_key_id = 'minio', secret_access_key = 'ClickHouse_Minio_******'). Stack trace:
E   
E   0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x00000000133b761f
E   1. DB::Exception::Exception(String&&, int, String, bool) @ 0x000000000c8623ce
E   2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000c861e80
E   3. DB::Exception::Exception<String const&, String&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<String&>::type>, String const&, String&) @ 0x000000001345696b
E   4. DB::SensitiveDataMasker::MaskingRule::apply(String&) const @ 0x0000000013456217
E   5. DB::wipeSensitiveDataAndCutToLength(String, unsigned long) @ 0x000000001345633c
E   6. DB::IAST::formatWithPossiblyHidingSensitiveData(unsigned long, bool, bool, bool, DB::IdentifierQuotingRule, DB::IdentifierQuotingStyle) const @ 0x000000001a65ee00
E   7. DB::InDepthNodeVisitor<DB::DiskConfigurationFlattener, false, false, std::shared_ptr<DB::IAST>>::doVisit(std::shared_ptr<DB::IAST>&) @ 0x000000001723d9a6
E   8. DB::DiskFromAST::createCustomDisk(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, bool) @ 0x000000001723ba4f
E   9. DB::MergeTreeSettingsImpl::loadFromQuery(DB::ASTStorage&, std::shared_ptr<DB::Context const>, bool) @ 0x00000000194863c2
E   10. DB::create(DB::StorageFactory::Arguments const&) @ 0x0000000019687efd
E   11. DB::StorageFactory::get(DB::ASTCreateQuery const&, String const&, std::shared_ptr<DB::Context>, std::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, DB::LoadingStrictnessLevel, bool) const @ 0x0000000018a3b6fb
E   12. DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::unique_ptr<DB::DDLGuard, std::default_delete<DB::DDLGuard>>&, DB::LoadingStrictnessLevel) @ 0x0000000017eedaf5
E   13. DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x0000000017ee22f6
E   14. DB::InterpreterCreateQuery::execute() @ 0x0000000017ef63f8
E   15. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, std::unique_ptr<DB::ReadBuffer, std::default_delete<DB::ReadBuffer>>&, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::ImplicitTransactionControlExecutor>) @ 0x0000000018385b92
E   16. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x000000001837de0b
E   17. DB::TCPHandler::runImpl() @ 0x0000000019a5a0ca
E   18. DB::TCPHandler::run() @ 0x0000000019a7c119
E   19. Poco::Net::TCPServerConnection::start() @ 0x000000001efa03c7
E   20. Poco::Net::TCPServerDispatcher::run() @ 0x000000001efa0859
E   21. Poco::PooledThread::run() @ 0x000000001ef66e47
E   22. Poco::ThreadImpl::runnableEntry(void*) @ 0x000000001ef65241
E   23. ? @ 0x0000000000094ac3
E   24. ? @ 0x00000000001268c0
```

--- a/src/Common/SensitiveDataMasker.cpp
+++ b/src/Common/SensitiveDataMasker.cpp
@@ -13,9 +13,8 @@
 #include <Common/StringUtils.h>
 #include <Common/ProfileEvents.h>
 
-#ifndef NDEBUG
-#    include <iostream>
-#endif
+
+#include <iostream>
 
 
 namespace ProfileEvents
@@ -67,6 +66,8 @@ public:
         , regexp(regexp_string, RE2::Quiet)
         , replacement(replacement_string)
     {
+        std::cout << "Loaded rule " << name << " (regexp: " << regexp_string << ", replacement: " << replacement_string << ", throw_on_match: " << throw_on_match << ")" << std::endl;
+
         if (!regexp.ok())
             throw Exception(ErrorCodes::CANNOT_COMPILE_REGEXP,
                 "SensitiveDataMasker: cannot compile re2: {}, error: {}. "
@@ -79,9 +80,14 @@ public:
         auto m = RE2::GlobalReplace(&data, regexp, replacement);
 
         if (throw_on_match && m > 0)
+        {
+            std::cout << "Trigggered " << regexp_string << " was triggered on the log line " << data << ", throw_on_match: {}" << throw_on_match << std::endl;
+
             throw Exception(ErrorCodes::LOGICAL_ERROR,
-                            "The rule {} was triggered on the log line {}",
-                            name, data);
+                "The rule {} was triggered on the log line {}",
+                name, data);
+        }
+
 
 #ifndef NDEBUG
         matches_count += m;
@@ -142,15 +148,22 @@ SensitiveDataMasker::SensitiveDataMasker(const Poco::Util::AbstractConfiguration
     LoggerPtr logger = getLogger("SensitiveDataMaskerConfigRead");
 
     std::set<std::string> used_names;
+    std::set<std::string> used_rules;
 
     for (const auto & rule : keys)
     {
+        /// Rules names are expected to be unique and be in a form of "rule1", "rule2", etc.
         if (startsWith(rule, "rule"))
         {
             auto rule_config_prefix = config_prefix + "." + rule;
 
-            auto rule_name = config.getString(rule_config_prefix + ".name", rule_config_prefix);
+            if (!used_rules.insert(rule).second)
+            {
+                throw Exception(ErrorCodes::INVALID_CONFIG_PARAMETER,
+                    "There are at least two rules with the same prefix '{}' in the query_masking_rules configuration", rule);
+            }
 
+            auto rule_name = config.getString(rule_config_prefix + ".name", rule_config_prefix);
             if (!used_names.insert(rule_name).second)
             {
                 throw Exception(ErrorCodes::INVALID_CONFIG_PARAMETER,

--- a/tests/config/config.d/query_masking_rules.xml
+++ b/tests/config/config.d/query_masking_rules.xml
@@ -1,14 +1,14 @@
 <!-- Config for test server -->
 <clickhouse>
     <query_masking_rules>
-        <rule>
+        <rule_topsecret>
             <regexp>TOPSECRET.TOPSECRET</regexp>
             <replace>[hidden]</replace>
-        </rule>
-        <rule>
+        </rule_topsecret>
+        <rule_passwords_in_tests>
             <name>Detect passwords in tests</name>
             <regexp>(?i)P@ssw0rd</regexp>
             <throw_on_match>true</throw_on_match>
-        </rule>
+        </rule_passwords_in_tests>
     </query_masking_rules>
 </clickhouse>

--- a/tests/integration/helpers/0_common_masking_rules.xml
+++ b/tests/integration/helpers/0_common_masking_rules.xml
@@ -1,1 +1,15 @@
-../../../tests/config/config.d/query_masking_rules.xml
+<!-- Config for test server -->
+<!-- Same file as tests/config/config.d/query_masking_rules.xml -->
+<clickhouse>
+    <query_masking_rules>
+        <rule>
+            <regexp>TOPSECRET.TOPSECRET</regexp>
+            <replace>[hidden]</replace>
+        </rule>
+        <rule>
+            <name>Detect passwords in tests</name>
+            <regexp>(?i)P@ssw0rd</regexp>
+            <throw_on_match>true</throw_on_match>
+        </rule>
+    </query_masking_rules>
+</clickhouse>

--- a/tests/integration/helpers/0_common_masking_rules.xml
+++ b/tests/integration/helpers/0_common_masking_rules.xml
@@ -2,10 +2,10 @@
 <!-- Same file as tests/config/config.d/query_masking_rules.xml -->
 <clickhouse>
     <query_masking_rules>
-        <rule>
+        <rule_topsecret>
             <name>Detect passwords in tests</name>
             <regexp>(?i)P@ssw0rd</regexp>
             <throw_on_match>true</throw_on_match>
-        </rule>
+        </rule_topsecret>
     </query_masking_rules>
 </clickhouse>

--- a/tests/integration/test_plain_rewritable_backward_compatibility/configs/remove_masking_rules.xml
+++ b/tests/integration/test_plain_rewritable_backward_compatibility/configs/remove_masking_rules.xml
@@ -1,0 +1,7 @@
+<clickhouse>
+    <!--
+        Remove all masking rules (enabled by default in every test).
+        This config will be enabled for the old instance only.
+    -->
+    <query_masking_rules remove="remove"></query_masking_rules>
+</clickhouse>

--- a/tests/integration/test_plain_rewritable_backward_compatibility/test.py
+++ b/tests/integration/test_plain_rewritable_backward_compatibility/test.py
@@ -16,10 +16,15 @@ setup_settings = {
     'stay_alive': True,
     'with_remote_database_disk': False,
 }
-node_25_4 = cluster.add_instance("node_25_4", image="clickhouse/clickhouse-server", tag="25.4", with_installed_binary=True, **setup_settings)
-node_25_6 = cluster.add_instance("node_25_6", image="clickhouse/clickhouse-server", tag="25.6", with_installed_binary=True, **setup_settings)
-node_25_8 = cluster.add_instance("node_25_8", image="clickhouse/clickhouse-server", tag="25.8", with_installed_binary=True, **setup_settings)
-node_25_10 = cluster.add_instance("node_25_10", image="clickhouse/clickhouse-server", tag="25.10", with_installed_binary=True, **setup_settings)
+
+main_configs_to_replace = {
+    'main_configs' : ["configs/storage_conf.xml", "configs/remove_masking_rules.xml"]
+}
+
+node_25_4 = cluster.add_instance("node_25_4", image="clickhouse/clickhouse-server", tag="25.4", with_installed_binary=True, **setup_settings | main_configs_to_replace)
+node_25_6 = cluster.add_instance("node_25_6", image="clickhouse/clickhouse-server", tag="25.6", with_installed_binary=True, **setup_settings | main_configs_to_replace)
+node_25_8 = cluster.add_instance("node_25_8", image="clickhouse/clickhouse-server", tag="25.8", with_installed_binary=True, **setup_settings | main_configs_to_replace)
+node_25_10 = cluster.add_instance("node_25_10", image="clickhouse/clickhouse-server", tag="25.10", with_installed_binary=True, **setup_settings | main_configs_to_replace)
 node_master = cluster.add_instance("node_master", **setup_settings)
 
 logger = logging.getLogger(__name__)

